### PR TITLE
Preallocated Array

### PR DESCRIPTION
**Pull request by Panke**
Installs a random-access container interface on top of a preallocated
random-access range.

### CODE REVIEW

**Comment by Panke** [comment]
This was FixedArray and Sorted ( #2793 )

**Comment by andralex** [comment]
Thanks! First you may want to mind the [autotester failures](https://auto-tester.puremagic.com/pull-history.ghtml?projectid=1&repoid=3&pullid=2939).

**Comment by Panke** [comment]
Can be reviewed again.

**Comment by burner** [comment]
I'm not really sure what this offers over build in arrays in combination with std.array.

**Comment by Panke** [comment]
@burner I created PreallocatedArray to factor out code that was used both in std.container.BinaryHeap and std.container.Sorted (#2793)

This could not be done with std.array.

**Comment by burner** [comment]
please eleborate

**Comment by Panke** [comment]
> I'm not really sure what this offers over build in arrays in combination with std.array.

If you want to use preallocated storage (e.g. stack space or what not) std.array does not help you at all, because you'll need to manage the maximum length (or capacity) and the current length by hand. It also does not call any destructors etc. because it treats your array as a range and not as a container.

PreallocatedArray encapsulates this and additionally accepts any random acces range or an std.container as storage backend. Take a look at std.container.BinaryHeap

https://github.com/D-Programming-Language/phobos/blob/master/std/container/binaryheap.d#L148
https://github.com/D-Programming-Language/phobos/blob/master/std/container/binaryheap.d#L261
https://github.com/D-Programming-Language/phobos/blob/master/std/container/binaryheap.d#L295

where the implementation depends on the backend. This was the same situation for std.container.Sorted, so I've wanted to factor that out.

File: std/experimental/preallocated_array.d [hunk 1]
```diff
@@ -0,0 +1,542 @@
+module std.experimental.preallocated_array;
+
+
+public import std.container.util;
+version(unittest) import std.stdio, std.container, std.range, std.algorithm;
+
+/**
+Array type that uses a $(XREF2 range, isRandomAccessRange, random access range)
+as fixed storage location. No memory is ever allocated. Throws if not enough
```
Comment (burner) [inline std/experimental/preallocated_array.d]: I believe memory is allocated:
Payload.ctor -> RefCounted.initialize -> malloc

File: std/experimental/preallocated_array.d [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+unittest 
+{
+    size_t[12] store;
+    auto fx = PreallocatedArray!(size_t[])(store[], 0);
+
+    foreach(i; 0 .. store.length)
+        fx.insertBack(i);
+    
+    foreach(i; 1 .. store.length)
+        assert(fx[i] > fx[i - 1]);
```
Comment (burner) [inline std/experimental/preallocated_array.d]: please add: assert(store[i] > store[i-1]);

File: std/experimental/preallocated_array.d [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+ */
+    alias Element = ElementType!Range;
+
+    private:
+
+    static struct Data 
+    {
+        Store store;
+        size_t length;
+    }
```
Comment (burner) [inline std/experimental/preallocated_array.d]: this seams over engineered. Just keep app pointer to the array and store the length

File: std/experimental/preallocated_array.d [hunk 4]
```diff
@@ -0,0 +1,542 @@
...
+
+static if (is(typeof(store.dup)))
+{
+/**
+Duplicates the container. The elements themselves are not transitively
+duplicated. The underlying store must support $(D .dup) as well.
+
+Complexity: $(BIGOH n).
+ */
+    @property PreallocatedArray dup()
```
Comment (burner) [inline std/experimental/preallocated_array.d]: that feels wrong. Why would you dup something that holds a reference to stack data.

I see bugs like this happening.
```d
auto fun() {
    ....
    PreallocatedArray a;
    return a.dup;
}
```

File: std/experimental/preallocated_array.d [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+        
+        size_t start = min(length, nlength);
+        size_t stop = max(length, nlength);
+        _payload.length = nlength;
+        
+        for(size_t idx = start; idx < stop; ++idx)
+        {
+            store[idx] = Element.init;
+        }
+    }
```
Comment (burner) [inline std/experimental/preallocated_array.d]: this will allocated again

File: std/experimental/preallocated_array.d [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+
+Complexity: $(BIGOH 1).
+ */
+    @property size_t length() const
+    {
+        return _payload is Payload.init ? 0 : _payload.length;
+    }
+
+    /// ditto
+    alias opDollar = length;
```
Comment (burner) [inline std/experimental/preallocated_array.d]: empty and length could be alias this if you didn't use RefCounted. Less code would be better here.

**Comment by burner** [comment]
normal arrays call destructors, not sure why you call this a new feature. please elaborate.

File: std/experimental/preallocated_array.d [hunk 7]
```diff
@@ -0,0 +1,542 @@
...
+Complexity: $(BIGOH m * log(n)), where $(D m) is the number of
+elements in $(D stuff)
+ */
+    size_t insertBack(Stuff)(Stuff stuff)
+        if (isImplicitlyConvertible!(Stuff, Element))
+    {
+        assert(_payload.refCountedStore.isInitialized);
+        if (length == capacity)
+            throw new RangeError("not enough room to insert stuff");
+        length = length + 1;
```
Comment (burner) [inline std/experimental/preallocated_array.d]: what will happen if you have a stack allocated array and you try to increase the size.

File: std/experimental/preallocated_array.d [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+ * Wrap store, which must be a $(XREF2 range, isRandomAccessRange, random access range)
+ * in a preallocated array and return it
+ */
+auto preallocatedArray(Store)(Store s, size_t initialLength = size_t.max)
+{
+	auto fA = PreallocatedArray!(Store)(s, initialLength);
+	return fA;
+}
+
+/// create a fixed array over  double[]
```
Comment (CyberShadow) [inline std/experimental/preallocated_array.d]: I think this example should be at the top.

**Comment by burner** [comment]
ping

**Comment by DmitryOlshansky** [comment]
@Panke Any plans to continue working on this? Going to close within a week otherwise.

**Comment by JakobOvrum** [comment]
Closing this due to inactivity. @Panke, please make a new PR or notify us if you continue work on this.

**Pull request closed without merging**